Teleportation/Tasks.qs

```diff
@@ -150,6 +150,29 @@ namespace Quantum.Kata.Teleportation {
 
         // ...
     }
+
+    // Task 1.8. Entanglement swapping
+    // Alice and Bob, independently from each other, each hold an entangled qubit pair in the 
+    // state |Φ⁺⟩ = (|00⟩ + |11⟩) / sqrt(2). They hand off one part of their pair to Charlie.
+    //
+    // Charlie can now teleport the state of Alice's qubit onto Bob's qubit, thus teleporting the entanglement. 
+    // Just like in "standard" teleportation, Bob still needs to apply the reconstruction steps -
+    // based on Charlie's measurement results - to the qubit still in his possession.
+    //
+    // The practical outcome is that the entangled state |Φ⁺⟩ = (|00⟩ + |11⟩) / sqrt(2) now spans across 
+    // Alice's and Bob's qubits - the ones that they didn't send to Charlie. They are now maximally entangled
+    // even though they never interacted in the first place!
+    //
+    // Goal: Entangle one part of Alice's input qubits pair with one part of Bob's input qubit pair,
```

Comment:
The way the task is formulated now, it's not clear which of Alice's qubits should end up entangled with which of Bob's. What is worse, the way the test is written, there is nothing preventing the solution from just swapping entangledPairAlice[0] with entangledPairBob[0], yielding not one but two pairs of entangled qubits. And it's pretty hard to verify in a test that two specific qubits never interact in a single multi-qubit gate... (This is exactly the same problem task 1.4 has, actually. We have tasks in which we prohibit all multi-qubit gates other than measurements, but that's slightly different)

This task is really about figuring out which of the building blocks of teleportation to apply at which location, since it doesn't use any new blocks. I'd make this task an open-ended one, similar to task 1.7, so that the learner has to figure the steps of the protocol, prepare the pairs, apply the right sequence of the previous tasks and figure out how to demonstrate it - but it's up to them to figure out which qubits they want to entangle in the end, and we don't have to develop a test to validate that the solution indeed performs the teleport :-)

Reply:
Thanks a lot for a quick and very helpful feedback. I now see your point of view when approaching these katas better.

I have reworked the kata to incorporate your suggestions. With the updated version, the reference implementation is only 2 lines, which is nice because all it boils down to is to reuse previous task solutions in a different context. Also with  the update version, the user no longer gets both EPR pairs but instead both Bob's qubits but only one part of the pair from Alice.

The task is then rephrased to use teleportation to create entanglement between Bob's 2nd qubit and Alice's 2nd qubit to - which the user is not even given access to. The only place that accesses Alice's 2nd qubit is the test, which still verifies the result. I hope this is now conveying the intent and the expectations for the user better and that it allows us to keep the test for feedback purposes. 

However, if you wish, I can rework this into an open-ended task (where everything, including the initial entanglement of both pairs, has to be done by the user) as well!

Reply:
As long as the code has access to one of the qubits from Alice's pair and both qubits from Bob's pair, the SWAP solution still passes - swap the qubit from Alice's pair with the one from Bob's pair that is supposed to be part of the final entangled state. To prevent this from happening, we need to give the code access to Charlie's qubits only, which becomes effectively task 1.2 again...

I wonder if we can implement this task in a more functional programming style - not give access to the qubits directly, but rather have a task which takes no arguments and returns two operations - one that Charlie should do (takes 2 qubits as inputs and returns an integer) and one that Bob should do (takes a qubit and an integer). This way we never give the learner's code access to the qubits they shouldn't have, so it doesn't have a way to do SWAP. The testing harness then would call these methods sequentially and check that as a result Alice's and Bob's qubits are in the |00>+|11> state. And the reference solution would combine reference solutions to tasks 1.2 and 1.3 with a wrapper to convert a tuple of Boolean values to a single integer and vice versa (or the learner can try and come up with something different).

What do you think about this approach?

Reply:
of course you are right 🤦‍♂️

I have changed the task to the functional style as suggested - it's a great idea. The goal is now to produce a pair of callables to be returned in a tuple, as suggested. The rest is pretty much the same as before.